### PR TITLE
refactor(web): system-settings

### PR DESCRIPTION
**Pull request by martabal**
## Changes made in this PR

- reduce the number of API requests
- improve the animations when toggling settings
- currently, clicking on the `Reset to default` button disappear only after clicking on `Save`. It now disappears right after clicking on it

## Screenshots

| Before | After |
| --- | --- |
| <video src="https://github.com/immich-app/immich/assets/74269598/d71aa86b-d68a-4d2b-ba74-5ddcc99b01b8"> | <video src="https://github.com/immich-app/immich/assets/74269598/c4828423-d728-423e-9e9b-b29f32dc1725">|

Edit: moved the data fetching on the page load from the client side to the server side so no more loading spinner !

### CODE REVIEW

**Review by jrasm91 — COMMENTED** [review]
It might require more work, but really how this should work (IMO) is the *Settings components don't make any API requests themselves. The all take in the current config (the entire object, not just the target section), show their relevant section and allow for modifications. Clicking save emits an event (`dispatch('save', newConfig)`) and the parent component has a single `handleSave` method that's re-used for all the `on:save` events of each child component. Each component can make a copy of the original configuration on mount to track local changes.

**Comment by martabal** [comment]
Using `bind:` so the `Unsaved change` label disappear only when the API request is successful.

File: web/src/routes/admin/system-settings/+page.svelte [hunk 1]
```diff
@@ -7,55 +7,152 @@
...
+    NotificationType,
+  } from '$lib/components/shared-components/notification/notification';
 
   export let data: PageData;
+  let config: SystemConfigDto = data.config;
 
-  const getConfig = async () => {
-    const { data } = await api.systemConfigApi.getConfig();
-    return data;
+  let currentConfig: SystemConfigDto = JSON.parse(JSON.stringify(config));
```
Comment (alextran1502) [inline web/src/routes/admin/system-settings/+page.svelte:20]: Why do we have to parse here? isn't the data coming back is already typed?

File: web/src/routes/admin/system-settings/+page.server.ts [hunk 2]
```diff
@@ -11,8 +12,17 @@ export const load: PageServerLoad = async ({ parent }) => {
     throw redirect(302, AppRoute.PHOTOS);
   }
 
+  const config: SystemConfigDto = await locals.api.systemConfigApi.getConfig().then((res) => res.data);
```
Comment (alextran1502) [inline web/src/routes/admin/system-settings/+page.server.ts:15]:
```suggestion
  const { data: config } = await locals.api.systemConfigApi.getConfig();
```

File: web/src/routes/admin/system-settings/+page.server.ts [hunk 3]
```diff
@@ -11,8 +12,17 @@ export const load: PageServerLoad = async ({ parent }) => {
...
+  const defaultConfig: SystemConfigDto = await locals.api.systemConfigApi.getDefaults().then((res) => res.data);
```
Comment (alextran1502) [inline web/src/routes/admin/system-settings/+page.server.ts:16]:
```suggestion
  const { data: defaultConfig } = await locals.api.systemConfigApi.getDefaults();
```

File: web/src/routes/admin/system-settings/+page.server.ts [hunk 4]
```diff
@@ -11,8 +12,17 @@ export const load: PageServerLoad = async ({ parent }) => {
...
+  const templateOptions: SystemConfigTemplateStorageOptionDto = await locals.api.systemConfigApi
+    .getStorageTemplateOptions()
+    .then((res) => res.data);
+
```
Comment (alextran1502) [inline web/src/routes/admin/system-settings/+page.server.ts:20]:
```suggestion
  const { data: templateOptions } = await locals.api.systemConfigApi
    .getStorageTemplateOptions();

```

File: web/src/routes/admin/system-settings/+page.svelte (hunk 1 above)
Comment (jrasm91) [reply web/src/routes/admin/system-settings/+page.svelte:20]: He needs an independent copy. If you assign it, both variables will reference the same object and both will be updated any time a property changes. This is a little hack to deep clone the object so you have two separate versions of it.

Comment (alextran1502) [reply web/src/routes/admin/system-settings/+page.svelte:20]: Ah got it, deep cloning makes sense

Comment (jrasm91) [reply web/src/routes/admin/system-settings/+page.svelte:20]: We already have lodash. Could use their deep clone instead.

Comment (martabal) [reply web/src/routes/admin/system-settings/+page.svelte]: Yes exactly !
Thanks for the loadash suggestion. I'm using it now

**Comment by jrasm91** [comment]
There are merge conflicts. There is also another configuration PR which will likely bring more conflicts. If you merge the conflicts now we can merge this one first. Or, we can wait and merge it after the config file one is merged.

**Comment by martabal** [comment]
@jrasm91 I'm currently on vacation, I'll be back next week. It's probably better to merge it after the config file one is merged.

**Comment by martabal** [comment]
Unless someone wants to take care of the merge

**Comment by jrasm91** [comment]
I think a lot of these changes are great, but there are too many merge conflicts now. Perhaps we can incrementally implement this. Some smaller steps could be:
- load defaults once server-side and pass it to all components
- load current config server-side and pass it to each setting as the initial config
- update each setting to emit events
- etc

**Pull request closed without merging**